### PR TITLE
Set/assign solver right before start of optimization

### PR DESCRIPTION
**Pull request by KeyDamo**
Please take a look at the comments associated with each commit.

### CODE REVIEW

File: solvers/NonlinearProgram.m [hunk 1]
```diff
@@ -155,7 +155,6 @@
       obj.bbcon_lb = [];
       obj.bbcon_ub = [];
       
-      obj = obj.setSolver('default');
```
Comment (avalenzu) [inline solvers/NonlinearProgram.m]: With this change we'd need to set the solver to 'default' in the constructor of every class that inherits from `NonlinearProgram`. What's your motivation for this change?

Comment (KeyDamo) [inline solvers/NonlinearProgram.m]: I don't see the need for doing that, or I didn't understand you.
Look at line 734 in NonlinearProgram.  Setting to 'default' causes setSolver to run through a series of if-elseif that look for a solver in some priority order, and in particular, for checking studentSnopt, the number of variables and constraints have to be already set.
You see the motivation in the comment I left on the commit. I'll paraphrase it here for you:
Run runDircolPerching.m having studentSnopt (and not the full Snopt) installed.  Make sure that N=41 or higher.  You'll understand when you see the error it throws.  I wrote it in the comment I left.  Hope this helps.

Comment (KeyDamo) [inline solvers/NonlinearProgram.m]: I see a problem that probably is related to what you were trying to tell me: that we won't be able to set "manually" the solver and will always be set to default before starting the optimization.  Ok, I'm thinking about a solution.

Comment (avalenzu) [inline solvers/NonlinearProgram.m]: I see where you're coming from. The point I'm making is a bit different though. By moving the `setSolver` call into `DirectTrajectoryOptimization` we'd no longer be calling it in any of the other classes that inherit from `NonlinearProgram` (eg. `InverseKinematics`). If we do want to move the `setSolver` call into `solve`, it should be `NonlinearProgram/solve` not `DirectTrajectoryOptimization/solve`.

Thanks for helping with our support for student snopt, by the way. Most of the Drake developers have regular snopt installed, which means the student version gets less testing.

Comment (KeyDamo) [inline solvers/NonlinearProgram.m]: My Pleasure.  Now I perfectly see your point.  I'm thinking about it.

Comment (KeyDamo) [inline solvers/NonlinearProgram.m]: @avalenzu You're right. I fixed it right as you suggested. Just pushed it. Thanks.

**Comment by RussTedrake** [comment]
just discussed this with @avalenzu and @hongkai-dai, the problem is that we want the call to solve() to be fast (there are many cases where we set up the problem once, but solve many times).  

of course you can handle this in your own code by calling `setSolver('default')` just before calling `solve()`. 

We think the best solution is to check inside `addConstraint` and `addDecisionVariable` for studentSnopt and to throw a more helpful warning/error (or reset the solver) there.  will open a new issue about it.

**Pull request closed without merging**